rollup/internal/controller/relayer/l2_relayer.go

```diff
@@ -1120,6 +1199,119 @@ func (r *Layer2Relayer) StopSenders() {
 	}
 }
 
+// fetchBlobFeeHistory returns the last WindowSec seconds of blob‐fee samples,
+// by reading L1Block table’s BlobBaseFee column.
+func (r *Layer2Relayer) fetchBlobFeeHistory(windowSec uint64) ([]*big.Int, error) {
+	latest, err := r.l1BlockOrm.GetLatestL1BlockHeight(r.ctx)
+	if err != nil {
+		return nil, fmt.Errorf("GetLatestL1BlockHeight: %w", err)
+	}
+	// bootstrap on first call
+	if r.lastFetchedBlock == 0 {
+		// start window
+		r.lastFetchedBlock = latest - windowSec/secondsPerBlock
+	}
+	from := r.lastFetchedBlock + 1
+	//if new blocks
+	if from <= latest {
+		raw, err := r.l1BlockOrm.GetBlobFeesInRange(r.ctx, from, latest)
+		if err != nil {
+			return nil, fmt.Errorf("GetBlobFeesInRange: %w", err)
+		}
+		// append them
+		for _, v := range raw {
+			r.feeHistory = append(r.feeHistory, new(big.Int).SetUint64(v))
+			r.lastFetchedBlock++
+		}
+	}
+
+	maxLen := int(windowSec / secondsPerBlock)
+	if len(r.feeHistory) > maxLen {
+		r.feeHistory = r.feeHistory[len(r.feeHistory)-maxLen:]
+	}
+
+	return r.feeHistory, nil
+}
+
+// calculateTargetPrice applies pct_min/ewma + relaxation to get a BigInt target
+func calculateTargetPrice(windowSec uint64, strategy StrategyParams, firstTime time.Time, history []*big.Int) *big.Int {
+	n := len(history)
+	if n == 0 {
+		return big.NewInt(0)
+	}
+	// convert to float64 Gwei
+	data := make([]float64, n)
```

Comment:
why do we copy from `[]big.Int` to `[]float64` here? Can't we run into precision and rounding issues then? Shouldn't it be possible to do all operations in `big.Int`? 

Also as described in [this comment](https://github.com/scroll-tech/scroll/pull/1659/files#r2091051562) we still sort the data array every time, which might be quite expensive. 

I think this is okay for now, we should address this in the next version though.

Reply:
The sort operation should be of negligible performance impact (we are talking about 7200 elements to sort through for a 24 hour period). Do you think it makes sense to invest time in removing this? 

Similar argument goes with float64. Dealing with big.Int will make for a more convoluted code when doing the exponent for the exponential and sigmoid relaxations. The loss of precision would be negligible (a couple wei), unless the blob base fee grows beyond unless the blob base fee grows to the order of millions of Gwei (i.e. ≳10⁶ Gwei, or ≳10¹⁵ Wei), at which point you’d still only lose a handful of Wei to floating-point imprecision in comparison. But I can still change all to big.Int, no problem.